storage/jellyfish_merkle/src/node_type/mod.rs

```diff
@@ -583,4 +646,47 @@ pub enum NodeDecodeError {
     /// The first byte of the input is not a known tag representing one of the variants.
     #[fail(display = "lead tag byte is unknown: {}", unknown_tag)]
     UnknownTag { unknown_tag: u8 },
+
+    /// No children found in internal node
+    #[fail(display = "No children found in internal node")]
+    NoChildren,
+
+    /// Extra leaf bits set
+    #[fail(
+        display = "Non-existent leaf bits set, existing: {}, leaves: {}",
+        existing, leaves
+    )]
+    ExtraLeaves { existing: u16, leaves: u16 },
+}
+
+/// Helper function to serialize version in a more efficient encoding.
+/// We use a super simple encoding - the high bit is set if more bytes follow.
+fn serialize_u64_varint(mut num: u64, binary: &mut Vec<u8>) -> Result<()> {
+    loop {
+        let low_bits = num as u8 & 0x7f;
+        num = num >> 7;
+        let more = (num > 0) as u8;
+        binary.push(low_bits | more << 7);
+        if more == 0 {
+            break;
+        }
+    }
+    Ok(())
+}
+
+/// Helper function to deserialize versions from above encoding.
+fn deserialize_u64_varint<T>(reader: &mut T) -> Result<u64>
+where
+    T: Read,
+{
+    let mut num = 0u64;
+    for _ in 0..=8 {
+        let byte = reader.read_u8()?;
+        let more = (byte & 0x80) != 0;
+        num = num << 7 | (byte & 0x7f) as u64;
+        if !more {
+            break;
+        }
+    }
+    Ok(num)
```

Comment:
... and `unreachable()` here?

Reply:
With a bad encoding, this is reachable.  Should I bother checking for and setting errors on bad encodings?

Reply:
I decided to check for bad encodings, better practice.